File: frontend/src/types/fsm.ts
// src/types/fsm.ts

export enum RuntimeState {
  IDLE = 'IDLE',
  THINKING = 'THINKING',
  PLANNING = 'PLANNING',
  EXECUTING = 'EXECUTING',
  VERIFYING = 'VERIFYING',
  RECOVERING = 'RECOVERING',
  FAILED = 'FAILED',
  COMPLETED = 'COMPLETED',
}

export type StateTransition = {
  from: RuntimeState | RuntimeState[];
  to: RuntimeState | RuntimeState[];
};

export const VALID_TRANSITIONS: StateTransition[] = [
  // Standard pipeline
  { from: RuntimeState.IDLE, to: RuntimeState.THINKING },
  { from: RuntimeState.THINKING, to: RuntimeState.PLANNING },
  { from: RuntimeState.THINKING, to: RuntimeState.EXECUTING }, // Direct tool call (no plan step)
  { from: RuntimeState.PLANNING, to: RuntimeState.EXECUTING },
  { from: RuntimeState.EXECUTING, to: RuntimeState.VERIFYING },
  { from: RuntimeState.EXECUTING, to: RuntimeState.FAILED },   // Immediate failure
  { from: RuntimeState.VERIFYING, to: [RuntimeState.EXECUTING, RuntimeState.RECOVERING, RuntimeState.COMPLETED] },
  { from: RuntimeState.RECOVERING, to: [RuntimeState.PLANNING, RuntimeState.EXECUTING, RuntimeState.FAILED] },

  // Terminal → reset
  { from: [RuntimeState.COMPLETED, RuntimeState.FAILED], to: RuntimeState.IDLE },

  // Emergency reset: any state can return to IDLE or FAILED (e.g. user abort, crash recovery, parsing error)
  { from: [RuntimeState.THINKING, RuntimeState.PLANNING, RuntimeState.EXECUTING, RuntimeState.VERIFYING, RuntimeState.RECOVERING], to: [RuntimeState.IDLE, RuntimeState.FAILED] },
];

File: src/aegis/core/fsm.py
from __future__ import annotations

from aegis.core.protocol import RuntimeState


VALID_TRANSITIONS: tuple[tuple[tuple[RuntimeState, ...], tuple[RuntimeState, ...]], ...] = (
    ((RuntimeState.IDLE,), (RuntimeState.THINKING,)),
    ((RuntimeState.THINKING,), (RuntimeState.PLANNING, RuntimeState.EXECUTING)),
    ((RuntimeState.PLANNING,), (RuntimeState.EXECUTING,)),
    ((RuntimeState.EXECUTING,), (RuntimeState.VERIFYING, RuntimeState.FAILED)),
    ((RuntimeState.VERIFYING,), (RuntimeState.EXECUTING, RuntimeState.RECOVERING, RuntimeState.COMPLETED)),
    ((RuntimeState.RECOVERING,), (RuntimeState.PLANNING, RuntimeState.EXECUTING, RuntimeState.FAILED)),
    ((RuntimeState.COMPLETED, RuntimeState.FAILED), (RuntimeState.IDLE,)),
    (
        (
            RuntimeState.THINKING,
            RuntimeState.PLANNING,
            RuntimeState.EXECUTING,
            RuntimeState.VERIFYING,
            RuntimeState.RECOVERING,
        ),
        (RuntimeState.IDLE, RuntimeState.FAILED),
    ),
)


def coerce_state(value: RuntimeState | str) -> RuntimeState:
    return value if isinstance(value, RuntimeState) else RuntimeState(value)


def is_valid_transition(from_state: RuntimeState | str, to_state: RuntimeState | str) -> bool:
    source = coerce_state(from_state)
    target = coerce_state(to_state)
    if source == target:
        return True
    return any(source in sources and target in targets for sources, targets in VALID_TRANSITIONS)
